sklearn/utils/_joblib.py

```diff
@@ -3,29 +3,18 @@
 import os as _os
 import warnings as _warnings
 
-# An environment variable to use the site joblib
-if _os.environ.get('SKLEARN_SITE_JOBLIB', False):
-    with _warnings.catch_warnings():
-        _warnings.simplefilter("ignore")
-        # joblib imports may raise DeprecationWarning on certain Python
-        # versions
-        import joblib
-        from joblib import logger
-        from joblib import dump, load
-        from joblib import __version__
-        from joblib import effective_n_jobs
-        from joblib import hash
-        from joblib import cpu_count, Parallel, Memory, delayed
-        from joblib import parallel_backend, register_parallel_backend
-else:
-    from ..externals import joblib
-    from ..externals.joblib import logger
-    from ..externals.joblib import dump, load
-    from ..externals.joblib import __version__
-    from ..externals.joblib import effective_n_jobs
-    from ..externals.joblib import hash
-    from ..externals.joblib import cpu_count, Parallel, Memory, delayed
-    from ..externals.joblib import parallel_backend, register_parallel_backend
+with _warnings.catch_warnings():
+    _warnings.simplefilter("ignore")
+    # joblib imports may raise DeprecationWarning on certain Python
+    # versions
+    import joblib
+    from joblib import logger
+    from joblib import dump, load
+    from joblib import __version__
+    from joblib import effective_n_jobs
+    from joblib import hash
+    from joblib import cpu_count, Parallel, Memory, delayed
+    from joblib import parallel_backend, register_parallel_backend
```

Comment:
We still keep the `sklearn.utils._joblib` module I suppose.

Reply:
Maybe @GaelVaroquaux @ogrisel @tomMoral can weight in: is joblib >= 0.11 still raising DeprecationWarnings on Python 3?

If not we could get rid of this file too.

Also @rth we can remove the comment at the top of the file mentionning local / site joblib

Reply:
@NicolasHug I agree we should probably get rid of `sklearn.utils._joblib` but that would involve updating a large number of imports

<details>

```
sklearn/multioutput.py
28:from .utils._joblib import Parallel, delayed

sklearn/pipeline.py
19:from .utils._joblib import Parallel, delayed

sklearn/multiclass.py
56:from .utils._joblib import Parallel
57:from .utils._joblib import delayed

sklearn/decomposition/dict_learning.py
16:from ..utils._joblib import Parallel, delayed, effective_n_jobs

sklearn/decomposition/online_lda.py
23:from ..utils._joblib import Parallel, delayed, effective_n_jobs

sklearn/datasets/rcv1.py
24:from ..utils import _joblib

sklearn/datasets/covtype.py
29:from ..utils import _joblib

sklearn/datasets/species_distributions.py
53:from sklearn.utils import _joblib

sklearn/datasets/twenty_newsgroups.py
47:from ..utils import _joblib

sklearn/datasets/olivetti_faces.py
26:from ..utils import _joblib

sklearn/datasets/kddcup99.py
24:from ..utils import _joblib

sklearn/datasets/california_housing.py
36:from ..utils import _joblib

sklearn/datasets/lfw.py
22:from ..utils._joblib import Memory
23:from ..utils import _joblib

sklearn/ensemble/partial_dependence.py
13:from ..utils._joblib import Parallel, delayed

sklearn/ensemble/voting_classifier.py
20:from ..utils._joblib import Parallel, delayed

sklearn/ensemble/iforest.py
18:from ..utils.fixes import _joblib_parallel_args

sklearn/ensemble/forest.py
52:from ..utils._joblib import Parallel, delayed
61:from ..utils.fixes import parallel_helper, _joblib_parallel_args

sklearn/ensemble/base.py
15:from ..utils._joblib import effective_n_jobs

sklearn/ensemble/bagging.py
15:from ..utils._joblib import Parallel, delayed

sklearn/linear_model/least_angle.py
24:from ..utils._joblib import Parallel, delayed

sklearn/linear_model/theil_sen.py
23:from ..utils._joblib import Parallel, delayed, effective_n_jobs

sklearn/linear_model/stochastic_gradient.py
12:from ..utils._joblib import Parallel, delayed
35:from ..utils.fixes import _joblib_parallel_args

sklearn/linear_model/coordinate_descent.py
21:from ..utils._joblib import Parallel, delayed, effective_n_jobs
23:from ..utils.fixes import _joblib_parallel_args

sklearn/linear_model/base.py
26:from ..utils._joblib import Parallel, delayed

sklearn/linear_model/omp.py
19:from ..utils._joblib import Parallel, delayed

sklearn/linear_model/logistic.py
36:from ..utils._joblib import Parallel, delayed, effective_n_jobs
37:from ..utils.fixes import _joblib_parallel_args

sklearn/neighbors/base.py
27:from ..utils._joblib import Parallel, delayed, effective_n_jobs
28:from ..utils._joblib import __version__ as joblib_version

sklearn/model_selection/_validation.py
25:from ..utils._joblib import Parallel, delayed
26:from ..utils._joblib import logger

sklearn/model_selection/_search.py
31:from ..utils._joblib import Parallel, delayed

sklearn/feature_selection/rfe.py
18:from ..utils._joblib import Parallel, delayed, effective_n_jobs

sklearn/metrics/pairwise.py
28:from ..utils._joblib import Parallel
29:from ..utils._joblib import delayed
30:from ..utils._joblib import effective_n_jobs

sklearn/utils/validation.py
26:from ._joblib import Memory
27:from ._joblib import __version__ as joblib_version

sklearn/utils/fixes.py
221:    from . import _joblib

sklearn/utils/__init__.py
15:from . import _joblib

sklearn/utils/testing.py
52:from sklearn.utils._joblib import joblib

sklearn/utils/estimator_checks.py
15:from sklearn.utils import _joblib

sklearn/tests/test_multioutput.py
20:from sklearn.utils._joblib import cpu_count

sklearn/tests/test_site_joblib.py
3:from sklearn.utils._joblib import Parallel, delayed, Memory, parallel_backend

sklearn/covariance/graph_lasso_.py
26:from ..utils._joblib import Parallel, delayed

sklearn/cluster/mean_shift_.py
26:from ..utils._joblib import Parallel
27:from ..utils._joblib import delayed

sklearn/tests/test_pipeline.py
34:from sklearn.utils._joblib import Memory
35:from sklearn.utils._joblib import __version__ as joblib_version

sklearn/cluster/k_means_.py
31:from ..utils._joblib import Parallel
32:from ..utils._joblib import delayed
33:from ..utils._joblib import effective_n_jobs

sklearn/compose/_column_transformer.py
17:from ..utils._joblib import Parallel, delayed

sklearn/manifold/mds.py
15:from ..utils._joblib import Parallel
16:from ..utils._joblib import delayed
17:from ..utils._joblib import effective_n_jobs

sklearn/ensemble/tests/test_bagging.py
36:from sklearn.utils import _joblib

sklearn/ensemble/tests/test_forest.py
25:from sklearn.utils._joblib import joblib
26:from sklearn.utils._joblib import parallel_backend
27:from sklearn.utils._joblib import register_parallel_backend
28:from sklearn.utils._joblib import __version__ as __joblib_version__

sklearn/linear_model/tests/test_sgd.py
30:from sklearn.utils import _joblib
31:from sklearn.utils._joblib import parallel_backend

sklearn/neighbors/tests/test_neighbors.py
29:from sklearn.utils._joblib import joblib
30:from sklearn.utils._joblib import parallel_backend

sklearn/neighbors/tests/test_kde.py
13:from sklearn.utils import _joblib

sklearn/metrics/tests/test_score_objects.py
40:from sklearn.utils import _joblib

sklearn/utils/tests/test_utils.py
306:    from sklearn.utils._joblib import Parallel, Memory, delayed
307:    from sklearn.utils._joblib import cpu_count, hash, effective_n_jobs
308:    from sklearn.utils._joblib import parallel_backend
309:    from sklearn.utils._joblib import register_parallel_backend
319:    from sklearn.utils._joblib import joblib

sklearn/utils/tests/test_estimator_checks.py
11:from sklearn.utils import _joblib
```

</details>

and I would rather do it in a separate PR, as this diff is large enough. It's a bit orthogonal to unvendoring anyway and rather how we alias or don't alias joblib imports. The points of this PR is to remove `sklearn.externals.joblib` with minimal changes elsewhere.

Removed the no longer relevant warning.

Reply:
I agree that the import should be changed in a separate PR to keep it readable.